cdc/kv/client.go

```diff
@@ -1064,64 +1073,119 @@ func (s *eventFeedSession) receiveFromStream(
+			if err != nil {
+				return err
 			}
+		}
+	}
+}
 
-			if isNewSubscription {
-				// It's the first response for this region. If the region is newly connected, the region info should
-				// have been put in `pendingRegions`. So here we load the region info from `pendingRegions` and start
-				// a new goroutine to handle messages from this region.
-				// Firstly load the region info.
-				state, ok = pendingRegions.take(event.RequestId)
-				if !ok {
-					log.Error("received an event but neither pending region nor running region was found",
-						zap.Uint64("regionID", event.RegionId),
-						zap.Uint64("requestID", event.RequestId),
-						zap.String("addr", addr))
-					return cerror.ErrNoPendingRegion.GenWithStackByArgs(event.RegionId, event.RequestId, addr)
-				}
+func (s *eventFeedSession) sendRegionChangeEvent(
+	ctx context.Context,
+	g *errgroup.Group,
+	event *cdcpb.Event,
+	regionStates map[uint64]*regionFeedState,
+	pendingRegions *syncRegionFeedStateMap,
+	addr string,
+	limiter *rate.Limiter,
+) error {
+	state, ok := regionStates[event.RegionId]
+	// Every region's range is locked before sending requests and unlocked after exiting, and the requestID
+	// is allocated while holding the range lock. Therefore the requestID is always incrementing. If a region
+	// is receiving messages with different requestID, only the messages with the larges requestID is valid.
+	isNewSubscription := !ok
+	if ok {
+		if state.requestID < event.RequestId {
+			log.Debug("region state entry will be replaced because received message of newer requestID",
+				zap.Uint64("regionID", event.RegionId),
+				zap.Uint64("oldRequestID", state.requestID),
+				zap.Uint64("requestID", event.RequestId),
+				zap.String("addr", addr))
+			isNewSubscription = true
+		} else if state.requestID > event.RequestId {
+			log.Warn("drop event due to event belongs to a stale request",
+				zap.Uint64("regionID", event.RegionId),
+				zap.Uint64("requestID", event.RequestId),
+				zap.Uint64("currRequestID", state.requestID),
+				zap.String("addr", addr))
+			return nil
+		}
+	}
 
-				// Then spawn the goroutine to process messages of this region.
-				regionStates[event.RegionId] = state
+	if isNewSubscription {
+		// It's the first response for this region. If the region is newly connected, the region info should
+		// have been put in `pendingRegions`. So here we load the region info from `pendingRegions` and start
+		// a new goroutine to handle messages from this region.
+		// Firstly load the region info.
+		state, ok = pendingRegions.take(event.RequestId)
+		if !ok {
+			log.Error("received an event but neither pending region nor running region was found",
+				zap.Uint64("regionID", event.RegionId),
+				zap.Uint64("requestID", event.RequestId),
+				zap.String("addr", addr))
+			return cerror.ErrNoPendingRegion.GenWithStackByArgs(event.RegionId, event.RequestId, addr)
+		}
```

Comment:
TiKV broadcasts resolved ts of all registered regions to every ticdc, so what if there are multiple ticdc in the same cluster?

Reply:
Oh, it is called by handling Event (not ResolvedTs), so it's ok.